Issue #26914 remove warning banner duplication

singhutsav5502 opened the pull request:
compose_banners: Add `visible_banners` set to prevent duplicate warning banners.
2 utility functions added for better readability,  `append_visible_banners` to append strings into visible_banners and `in_visible_banners` to check if a particular data string already exists in the set.

Introduces a "visible_banners" set in compose_banners to store strings containing details of currently rendered banners. This set is cleared in instances where `compose_actions.cancel()` is triggered, ensuring smooth functionality with hotkeys and the close button. The implementation checks for existing banners to prevent duplicate warning banners from being displayed when mentioning private streams. Duplicate warning banners are not created for the same stream while warnings for other streams still get rendered.


Fixes: https://github.com/zulip/zulip/issues/26914



![image](https://github.com/zulip/zulip/assets/29460617/5f2d68a9-8b4b-4283-9e1d-ed31be680d73)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

karlstolley commented on web/src/compose_banner.ts:
```diff
@@ -52,6 +52,8 @@ export const CLASSNAMES = {
     user_not_subscribed: "user_not_subscribed",
 };
 
+export const existing_banners = new Set();
```
I think this code will read just a bit more easily if you name this set `visible_banners`.

karlstolley commented:
Okay @singhutsav5502 -- note the naming comment above. There's been some significant refactoring in this area, so you'll need to carefully rebase your work here on `main`. Make sure you're also following our commit style documented at https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html

singhutsav5502 commented:
okaym i will rename it and look into rebasing my work without merge conflicts.

alya commented:
Please update your commit message to match the [commit style guidelines](https://zulip.readthedocs.io/en/latest/contributing/version-control.html#commit-messages), and post a comment when you've addressed all the feedback an believe your work is ready for review.

alya commented:
You will also need to re-add an fill out the self-review checklist in the PR template.

singhutsav5502 commented:
Hello, I have addressed  the naming issues of the variable, rebased my code and updated the commit message to make sure it adheres to the commit style guideline. I have also re-added the self review checklist and checked off all the relevant boxes and feel that my work should be ready for review. Please do inform me if there are any changes that i should make. Thanks.

alya commented:
Thanks! @karlstolley do you want to give this another round of review? I haven't tested the PR.

karlstolley commented on web/src/compose_banner.ts:
```diff
@@ -78,12 +81,21 @@ export function update_or_append_banner(
 export function clear_message_sent_banners(include_unmute_banner = true): void {
     for (const classname of Object.values(MESSAGE_SENT_CLASSNAMES)) {
-        $(`#compose_banners .${CSS.escape(classname)}`).remove();
+        const $banners = $(`#compose_banners .${CSS.escape(classname)}`);
```
I thought I commented on this earlier, but maybe it was in a different PR? I don't think reaching into the DOM is the right approach here. Why not make better use of the `visible_banners` Set and have that be the source of truth for the state of banners in the compose box? I see things get added, deleted, and cleared from that set here, but no where does it seem like that set has any bearing on the banners that are actually being displayed? Am I missing something?

singhutsav5502 replied:
https://github.com/zulip/zulip/blob/2a8ecd02cac3e2e06f733d0a710591fa97a66bf5/web/src/compose_validate.js#L132-L134

so far the `visible_banners` set has been used over here to check if the banner already exists or not . currently the set stores HTMLStrings instead of an object that consists of things like `classnames` which seems to be needed here as the function itself doesn't seem to accept a particular banner string to be deleted. so instead i ended up going through the elements that are getting selected one by one and clearing their strings from the set individually.

singhutsav5502 replied:
as an alternative instead of removing each banner individually i could simply clear the set but that would potentially lead to bugs since the `visible_banners` is container for all types of banners ( not just warning banners ).

karlstolley replied:
So why not store that information as part of an entry in the set?

singhutsav5502 replied:
the ` append_compose_banner_to_banner_list`  where i am adding the banners to the set function seems to have access to only the HTMLString and its container as the parameters. I will probably have to add more arguments to the function to get access to `banner type` and `stream name` which could lead to fundamental code changes throughout the compose banner area since this function seems to be used to add the banners at all places.

singhutsav5502 commented:
@karlstolley Hello, thanks for the review so far. I have updated the code accordingly based on your suggestions and made sure unnecessary DOM calls are being avoided. Please do inform me if any changes needed to be made.

singhutsav5502 commented:
gentle bump @alya @karlstolley , please review when possible, also there seems to be an issue regarding the claiming of the issue that this PR fixes, please look into it as well.

karlstolley commented:
I'm finding your code really difficult to follow and review. I don't understand why you're bouncing in and out of JSON. It complicates matters that you have put all of this work into a single commit, and that there are no comments in the code explaining what you're doing, or why.
